### DISCUSSION TITLE
Generating schemas from instances

### DISCUSSION THREAD

**Discussion by apirogov** [Materials-Data-Science-and-Informatics/metador-core#60] [Ideas]
After completing #59, it could be checked out how well JSON Schemas can be generated from example instances, using

https://pypi.org/project/genson/

or a similar library, and transforming them into metador schemas (at least "drafts").

**Comment by broeder-j** [comment]
Is there currently a way to do the 'transforming' from json schema/json-LD to metador schemas? because that will be the first step.
If one has this the reuse of existing schemas will be easier.

also here is a list of tools available for the schema generation from instances:
https://stackoverflow.com/questions/7341537/tool-to-generate-json-schema-from-json-data

**Reply by apirogov** [reply]
This is a general tool for pydantic: https://pydantic-docs.helpmanual.io/datamodel_code_generator/

Metador is based on pydantic, so I would expect it's not too hard to adapt. I have not tried it though, so one subtask would be to throw schemas at it and see what pydantic models it produces.

Mapping JSON-LD to pydantic/Metador automatically I would say is even more challenging. Someone tried this: https://github.com/lexiq-legal/pydantic_schemaorg

I would not call that stable and that would not immediately work with metador, but it is a relevant proof of concept of generating models from JSON-LD.

**Comment by broeder-j** [comment]
These might be useful schemas to allow:

1. For DFT at least: https://nomad-lab.eu/prod/rae/api/metainfo/ The experimental one seem very basic so far.
2. Codemeta for code.
3. Citation.cff

...
